### PR TITLE
Fix Gradient and Color Picker checkerboard, Fix tile parameter for CanvasItem.draw_texture_rect

### LINKED ISSUE #41561: Gradient editor displays bright magenta background instead of checker background for transparent colors
**Godot version:** Git https://github.com/godotengine/godot/commit/7a62b94bcae4da685144ce11aa6f1c8ad19b7c71

**OS/device including version:** Fedora 31, GeForce GTX 1080 (NVIDIA 440.100)

**Issue description:**

The Gradient editor fails to display its checker background for transparent colors and will display a bright magenta color instead.

image

This GUI component is drawn by [scene/gui/gradient_edit.cpp](https://github.com/godotengine/godot/blob/master/scene/gui/gradient_edit.cpp).

**Steps to reproduce:** Create a Gradient resource in the Inspector dock by using the New Resource button, then edit one of its points to have an alpha value lower than 1.0.

**Minimal reproduction project:** N/A

### LINKED ISSUE #42662: Vulkan: `CanvasItem.draw_texture_rect()` tiling does not work as expected
**Godot version:**
current master
3.x unaffected

**OS/device including version:**
Debian 10
Vulkan, Mesa DRI Intel(R) UHD Graphics 630

**Issue description:**
calling `CanvasItem::draw_texture_rect` with `tile = true` does not work or only in an unintuitive way. It definitive does not work as in 3.x

In 3.x this argument, well, tiled the texture.
In 4.0 it only prevents texture scaling but instead of tiling the edges just get stretched out.

Behavior in 3.2
image
left: TextureRect Stretch Mode tile, just a hidden call to draw_texture_rect really
right: manual call to draw_texture_rect from _draw()

The same setup in 4.0.dev
image

To achieve the same results in 4.0 one has to
- call `draw_texture_rect` with `tile = true` to prevent scaling instead of tiling
- and specify a tiling mode (calling `set_texture_repeat` or specifying the `texture_repeat argument`)

**Related Issues**
- reason for #40011
- background checkerboard does not correctly tile in ColorPicker related Nodes with alpha < 1: (example with alpha=0)
3.2:  image
4.0:  image
- possible that I missed some

**Possible Solutions**
- setting texture repeat as described above. some places in the engine already do this.
- changing the api to be less confusing. Maybe rename tile to keep_src_size as this currently describes the behavior, however compat breakage might be hard to spot as the same calls work but they don't do their job properly. 

**Steps to reproduce:**
any call to draw_texture_rect with tile=true

**Minimal reproduction project:**
[draw_texture_rect-tile.zip](https://github.com/godotengine/godot/files/5353415/draw_texture_rect-tile.zip) - setup as above

### PR DESCRIPTION
**Pull request by nathanfranke**
Fix #41561 
Fix #42662 

image

### CODE REVIEW

File: editor/editor_themes.cpp [hunk 1]
```diff
@@ -1144,11 +1144,8 @@ Ref<Theme> create_editor_theme(const Ref<Theme> p_theme) {
 	theme->set_constant("label_width", "ColorPicker", 10 * EDSCALE);
 	theme->set_icon("screen_picker", "ColorPicker", theme->get_icon("ColorPick", "EditorIcons"));
 	theme->set_icon("add_preset", "ColorPicker", theme->get_icon("Add", "EditorIcons"));
-	theme->set_icon("preset_bg", "ColorPicker", theme->get_icon("GuiMiniCheckerboard", "EditorIcons"));
```
Comment (nathanfranke) [inline editor/editor_themes.cpp:1147]: The icons "ColorPicker":"preset_bg" and "ColorPickerButton":"bg" are just references to the icon "EditorIcons":"GuiMiniCheckerboard". I'm removing the middle-man for now.

File: scene/gui/gradient_edit.cpp [hunk 2]
```diff
@@ -54,6 +54,9 @@ GradientEdit::GradientEdit() {
 
 	checker = Ref<ImageTexture>(memnew(ImageTexture));
 	Ref<Image> img = memnew(Image(checker_bg_png));
+	checker->create_from_image(img);
+	set_texture_repeat(TEXTURE_REPEAT_ENABLED); // Repeat background checkerboard
+	set_texture_filter(TEXTURE_FILTER_NEAREST); // Do not blur background checkerboard with filter
```
Comment (nathanfranke) [inline scene/gui/gradient_edit.cpp:59]: I wanted to use a `tile` parameter, but that doesn't exist on `draw_colored_polygon`

**Comment by nathanfranke** [comment]
Why is the checkerboard on the gradient ½ the size? Just using `draw_texture_rect`.

image

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp [hunk 3]
```diff
@@ -455,7 +455,7 @@ void RasterizerCanvasRD::_render_item(RD::DrawListID p_draw_list, const Item *p_
 
 				//bind textures
 
-				_bind_canvas_texture(p_draw_list, rect->texture, current_filter, current_repeat, last_texture, push_constant, texpixel_size);
+				_bind_canvas_texture(p_draw_list, rect->texture, current_filter, (rect->flags & CANVAS_RECT_TILE) ? RenderingServer::CanvasItemTextureRepeat::CANVAS_ITEM_TEXTURE_REPEAT_ENABLED : current_repeat, last_texture, push_constant, texpixel_size);
```
Comment (reduz) [inline servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:458]: this is probably unnecesary and the tile flag should be deprecated as you can change tiling in other means.

**Review by reduz — DISMISSED** [review]

File: servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp (hunk 3 above)
Comment (nathanfranke) [reply servers/rendering/rasterizer_rd/rasterizer_canvas_rd.cpp:458]: What should it be replaced with? Should repeat be a boolean in `CommandRect`?

**Comment by nathanfranke** [comment]
CC

> this is probably unnecesary and the tile flag should be deprecated as you can change tiling in other means.

What is the best way to change tile flags for editor icons? Also, maybe this can be removed in another PR.

File: editor/editor_themes.cpp (hunk 1 above)
Comment (KoBeWi) [reply editor/editor_themes.cpp:1347]: You can't remove them. It breaks ColorPickers used outside editor, they don't have access to editor icons. And also this makes them impossible to customize by the users.

**Comment by KoBeWi** [comment]
Looks ok, except the removal of ColorPicker icons (check my previous comment). Also needs rebase and SNAMEs.

File: editor/editor_themes.cpp (hunk 1 above)
Comment (nathanfranke) [reply editor/editor_themes.cpp]: Added them back. Also added the same thing for GradientEdit since previously it was hard coded in :)

Comment (KoBeWi) [reply editor/editor_themes.cpp]: GradientEdit doesn't need it though, because it's not exposed class.

**Review by KoBeWi — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by KoBeWi** [comment]
Uh I just realized that I forgot to check if you added the SNAMEs. The icons should use
`get_theme_icon(SNAME("GuiMiniCheckerboard"), SNAME("EditorIcons"))`

**Comment by nathanfranke** [comment]
Should I make a follow-up or can someone just make a commit for it?

**Comment by KoBeWi** [comment]
You can make a follow-up. Someone needs to fix it anyways.

**Pull request merged**